// @orxe-devkit/jest-config/src/jest.config.js
module.exports = function setupJest(customConfig) {
	const defaultConfig = {
		roots: ['<rootDir>/src'],
		collectCoverageFrom: [
			'<rootDir>/src/**/*.{js,jsx,ts,tsx}',
			'!<rootDir>/src/**/*.d.ts',
			'!<rootDir>/src/**/*css.ts',
		],
		testMatch: [
			'<rootDir>/src/**/__tests__/**/*.{js,jsx,ts,tsx}',
			'<rootDir>/src/**/*.{spec,test,axe}.{js,jsx,ts,tsx}',
		],
		runner: 'jest-runner',
		testEnvironment: 'jest-environment-jsdom-sixteen',
		transform: {
			'^.+\\.(js|jsx|ts|tsx)$': ['babel-jest', { rootMode: 'upward' }],
		},
		transformIgnorePatterns: ['^.+\\.(css|sass|scss)$'],
		moduleNameMapper: {
			'^.+\\.module\\.(css|sass|scss)$': 'identity-obj-proxy',
		},
		moduleFileExtensions: [
			'web.js',
			'js',
			'web.ts',
			'ts',
			'web.tsx',
			'tsx',
			'json',
			'web.jsx',
			'jsx',
			'node',
		],
		watchPlugins: ['jest-watch-typeahead/filename', 'jest-watch-typeahead/testname'],
		resolutions: {
			'babel-core': '^7.0.0-bridge.0',
		},
	};

	return { ...defaultConfig, ...customConfig };
};
