// @orxe-devkit/lint/eslint.js
module.exports = {
  extends: [
    'eslint:recommended',
    'google',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:wc/recommended',
  ],
  globals: {
    goog: false,
  },
  env: {
    browser: true,
  },
  parser: '@typescript-eslint/parser',
  plugins: ['@typescript-eslint'],
  parserOptions: {
    ecmaVersion: 2017,
    sourceType: 'module',
  },
  settings: {
    wc: {
      elementBaseClasses: ['BaseElement', 'LitElement'],
    },
  },
  rules: {
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/no-empty-interface': 'off',
    '@typescript-eslint/indent': 'off',
    'eol-last': 'off',
    quotes: 'off',
    'quote-props': 'off',
    indent: 'off',
    'no-console': 'warn',
    'new-cap': 'off',
    'no-tabs': 'off',
    'no-unused-vars': 'off',
    'no-invalid-this': 'off',
    'object-curly-spacing': 'off',
    '@typescript-eslint/no-non-null-assertion': 'off',
    'max-len': 'off',
    '@typescript-eslint/explicit-member-accessibility': [
      'error',
      {
        accessibility: 'no-public',
      },
    ],
    'no-new': 'warn',
    'no-var': 'error',
    'no-floating-decimal': 'error',
    'require-jsdoc': 'off',
    'valid-jsdoc': 'off',
    'prefer-const': 'error',
    'comma-dangle': 'off',
  },
  overrides: [
    {
      files: ['**/*.js'],
      rules: {
        '@typescript-eslint/no-var-requires': 'off',
        'no-undef': 'off',
      },
    },
  ],
};
